Absolute positioned checkboxes  overlay floated elements

rafh opened the pull request:
Currently, checkboxes are positioned as absolute. This positioning causes the input to overlay an element that has been floated within the editor. Floated elements are useful if you want your text to wrap around this element. This PR fixes the overlaying of checkboxes by removing the absolute positioning,  updating the `ul` padding, and displaying`.task-list-item` `flex` to ensure inputs and the associated label are on the same line.

Screenshots:

Before:
<img width="762" alt="Screenshot 2023-09-01 at 3 40 59 PM" src="https://github.com/go-gitea/gitea/assets/6152817/570247c7-7f5c-4697-bfc9-ad4655e37991">

After:
<img width="762" alt="Screenshot 2023-09-01 at 3 42 20 PM" src="https://github.com/go-gitea/gitea/assets/6152817/db53df45-1294-4eee-84c0-b21ac4fdf805">

silverwind commented:
We need to be careful around markdown rendering to stay compatible with GitHub's rendering. Could you post your sample markdown document here so we can compare this and GitHub's solution (assuming they have fixed it).

rafh commented:
Here's an example of my markdown. When an image is aligned left or right the following CSS is applied by GitHub 
```CSS
img[Attributes Style] {
    float: left;
    vertical-align: top;
    width: 200px;
}
```

<img align="left" width="200" src="/assets/6152817/d0b0b99a-db2b-4b6d-946d-277bfd24fc11" />

- [ ]  example checkbox

kdumontnu reviewed (approved):
For traceability, it looks likes this regression was introduced in https://github.com/go-gitea/gitea/issues/17395

I added a bit of a stress-test to compare and it looks like a marked improvement:
```
Here is some base text
- [ ] One
- [ ] 

<img align="left" width="300" height="200" src="https://gitea.com/assets/img/logo.svg">

- [ ] Two
- [ ] Three
- [ ] Four
- [ ] 

More text


#### dense list
- [ ] foo
- [ ] barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbar barbarbarbarbar
- [ ] baz

#### not so dense list
- [ ] foo

- [ ] barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbar barbarbarbarbar
- [ ] second item

- [ ] baz
```
![image](https://github.com/go-gitea/gitea/assets/12700993/8600d6eb-7044-4581-822f-c5c3331d7b3e)

rafh commented:
> Here's an example of my markdown. When an image is aligned left or right the following CSS is applied by GitHub
> 
> ```css
> img[Attributes Style] {
>     float: left;
>     vertical-align: top;
>     width: 200px;
> }
> ```
> 
> <img alt="" width="200" src="/assets/6152817/d0b0b99a-db2b-4b6d-946d-277bfd24fc11">
> 
> * [ ]   example checkbox

These associated markdown for this comment:

```MD
<img alt="" width="200" src="/assets/6152817/d0b0b99a-db2b-4b6d-946d-277bfd24fc11">
* [ ]   example checkbox
```

kdumontnu commented on web_src/css/markup/content.css:
```diff
@@ -144,7 +144,7 @@
 .markup ul,
 .markup ol {
-  padding-left: 2em;
+  padding-left: 0;
```
@rafh I'm worried about this style affecting other elements unrelated to checkboxes. Are there other `ul` `ol` items this will apply to in markdown (I assume yes)?

Especially because this value was not changed here: https://github.com/go-gitea/gitea/pull/17411/files

(btw, this is an area front-end testing could validate)

rafh replied:
@kdumontnu I understand your point. Instead, we should change the `em` back to `2em` and update the `.task-list-item` to have `margin-left: -2em;`. This will ensure that all checkboxes are aligned properly without impacting all `ol` and `ul`.

Here's an example of the markup and how it is rendered within Gitea.

```MD
- [x] example content

<img  align="left" alt="" width="200" src="https://user-images.githubusercontent.com/6152817/265531274-d0b0b99a-db2b-4b6d-946d-277bfd24fc11.jpg">

- [x] example content

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur sed efficitur orci, eu cursus sapien. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur sed efficitur orci, eu cursus sapien. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur sed efficitur orci, eu cursus sapien. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur sed efficitur orci, eu cursus sapien. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur sed efficitur orci, eu cursus sapien.

<img style="clear: both" alt="" width="200" src="https://user-images.githubusercontent.com/6152817/265531274-d0b0b99a-db2b-4b6d-946d-277bfd24fc11.jpg">


- [ ] one

- [ ] two
```

<img width="783" alt="Screenshot 2023-09-07 at 2 51 27 PM" src="https://github.com/go-gitea/gitea/assets/6152817/e93397e7-696c-4558-b84d-c99e427bf1e4">

kdumontnu replied:
That seems like a reasonable solution to me, or maybe you can add a `:not(.task-list-item)` to the selectors above and. Whatever will apply the style to the `.task-list-item` items, but not change the styling for other lists.

kdumontnu reviewed (approved):
Looks good - one minor thing below you might want to address is that the spacing between non-dense checkboxes is greater than spacing between non-dense list items (code below)

<img width="352" alt="Screenshot 2023-09-12 at 2 28 57 PM" src="https://github.com/go-gitea/gitea/assets/12700993/50f98a67-63f0-46d4-949b-4c39231bea30">

```
Here is some base text
- [ ] One
- [ ] 

- Bullet
- Items

<img align="left" width="300" height="200" src="https://gitea.com/assets/img/logo.svg">

- [ ] Two
- [ ] Three
- [ ] Four
- [ ] 

More text


#### dense list
- [ ] foo
- [ ] barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbar barbarbarbarbar
- [ ] baz

#### not so dense list
- [ ] foo

- [ ] barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbar barbarbarbarbar
- [ ] second item

- [ ] baz

<img align="left" width="300" height="200" src="https://gitea.com/assets/img/logo.svg">

- bullet
- items
```

Here is some base text
- [ ] One
- [ ] 

- Bullet
- Items

<img align="left" width="300" height="200" src="https://gitea.com/assets/img/logo.svg">

- [ ] Two
- [ ] Three
- [ ] Four
- [ ] 

More text


#### dense list
- [ ] foo
- [ ] barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbar barbarbarbarbar
- [ ] baz

#### not so dense list
- [ ] foo

- [ ] barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbar barbarbarbarbar
- [ ] second item

- [ ] baz

<img align="left" width="300" height="200" src="https://gitea.com/assets/img/logo.svg">

- bullet
- items

kdumontnu commented:
@silverwind any other thoughts on this - seems like a clear improvement, and gets rid of the absolute positioning (which I'm sure is messing up some other rendering as well).

rafh commented:
This is how GitHub renders the following markdown and how Gitea renders

**Gitea(left) - GitHub(right)**
![Group 19](https://github.com/go-gitea/gitea/assets/6152817/9cd6bafe-6850-4140-9de9-166471f81348)



```Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur

- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur

- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur
 <img align="left" width="300" height="200" src="https://github.com/assets/6152817/d0b0b99a-db2b-4b6d-946d-277bfd24fc11">

- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur

#### dense list
- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur
- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur
- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur

> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur

- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur

 <img align="left" width="300" height="200" src="https://github.com/assets/6152817/d0b0b99a-db2b-4b6d-946d-277bfd24fc11">

#### not so dense list
- [ ] 1
- [ ] 2
- [ ] 3```

delvh reviewed (approved):
I trust @kdumontnu's words that all potential regressions have been fixed.

wxiaoguang commented:
I do no think this fix is right. 

It breaks some more common layout like this:

```
* [ ] a
    * [ ] a1
    * [ ] a2
* [ ] b
* [ ] c
* [ ] d

----

* [ ] a
    * [ ] a1
    * [ ] a2

* [ ] b

```


![image](https://github.com/go-gitea/gitea/assets/2114189/06493ec1-bde7-4871-84d3-030ab5830b59)

kdumontnu commented:
> I do no think this fix is right.
> 
> It breaks some more common layout like this:
> 
> ```
> * [ ] a
>     * [ ] a1
>     * [ ] a2
> * [ ] b
> * [ ] c
> * [ ] d
> 
> ----
> 
> * [ ] a
>     * [ ] a1
>     * [ ] a2
> 
> * [ ] b
> ```

Thanks, good catch!

silverwind commented:
Can we as a first step copy GitHub's CSS exactly, and then improve upon it by adding the margin-left to the checkboxes so they are not "glued" to the image?

rafh commented:
> Can we as a first step copy GitHub's CSS exactly to reproduce their rendering, and then improve upon it by adding the margin-left to the checkboxes so they are not "glued" to the image?

I've applied the styling used by GitHub. I added more padding to left-aligned images. Otherwise, our checkboxes would lie directly next to the images. 

Below I've consolidated all the markdown examples and added one more example at the top (empty checkboxes). I added this because the previous code had the following comment and I wanted to ensure this edge case was covered.

> to render a checkbox list without content `- [ ]`, we need this min-height to make sure the `<li>` can be visible

```
Empty Checkboxes
- [ ] 
- [ ] 

--- 

* [ ] a
    * [ ] a1
    * [ ] a2
* [ ] b
* [ ] c
* [ ] d

----

* [ ] a
    * [ ] a1
    * [ ] a2

* [ ] b


- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur

- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur

 <img align="left" width="300" height="200" src="https://github.com/assets/6152817/d0b0b99a-db2b-4b6d-946d-277bfd24fc11">

- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur

#### dense list
- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur
- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur
- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur

> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur

- [ ] Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed aliquam velit ac metus vulputate egestas. Etiam ut enim iaculis, egestas mauris ac, convallis ex. Curabitur

 <img align="left" width="300" height="200" src="https://github.com/assets/6152817/d0b0b99a-db2b-4b6d-946d-277bfd24fc11">

#### not so dense list
- [ ] 1
- [ ] 2
- [ ] 3


Here is some base text
- [ ] One
- [ ] 

- Bullet
- Items

<img align="left" width="300" height="200" src="https://gitea.com/assets/img/logo.svg">

- [ ] Two
- [ ] Three
- [ ] Four
- [ ] 

More text


#### dense list
- [ ] foo
- [ ] barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbar barbarbarbarbar
- [ ] baz

#### not so dense list
- [ ] foo

- [ ] barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbarbar barbarbarbarbarbarbar barbarbarbarbar
- [ ] second item

- [ ] baz

<img align="left" width="300" height="200" src="https://gitea.com/assets/img/logo.svg">

- bullet
- items
```
